std/algorithm/sorting.d

```diff
@@ -2625,11 +2625,21 @@ private template TimSortImpl(alias pred, R)
             // can't use `temp.length` if there's no default constructor
             static if (__traits(compiles, { T defaultConstructed; cast(void) defaultConstructed; }))
             {
-                if (__ctfe) temp.length = newSize;
-                else temp = () @trusted { return uninitializedArray!(T[])(newSize); }();
+
+                static if (hasElaborateAssign!T)
+                    temp.length = newSize;
+                else
+                {
+                    if (__ctfe) temp.length = newSize;
+                    else temp = () @trusted { return uninitializedArray!(T[])(newSize); }();
+                }
             }
             else
             {
+                static assert(!hasElaborateAssign!T,
+                              "Structs which have opAssign but cannot be default-initialized " ~
+                              "do not currently work with stable sort: " ~
+                              "https://issues.dlang.org/show_bug.cgi?id=24810");
```

Comment:
this says "issue 24810", but this PR fixes issue 24809, the two issues look identical to me, are they dupes of each other?

Reply:
No, they are not identical.

24809 is about stable sort using uninitialized memory with types that have `opAssign`.

24810 is about stable sort using uninitialized memory with types that have `opAssign` _and_ which disable default initialization.

I specifically separated out the bug reports, because fixing the case with `opAssign` is easy, but fixing the case where the type both defines `opAssign` and disables default initialization is far more complicated. So, you could view this as a partial fix of the overall problem, but it fixes one of the cases while leaving the other one broken for the moment. The static assertion is so that if other code within `TimSortImpl` gets fixed so that it compiles with types that both have `opAssign` and disable default initialization, this part of the code won't start compiling and doing the wrong thing.

Reply:
Noted.